File: neetcode/125. Valid Palindrome/125.java
/*
Method 1.0: Two pointers
TC: O(N)
SC: O(1)
 */

class Solution {
    public boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;

        s = s.toLowerCase();

        while (left < right) {
            while (left < right && !Character.isLetterOrDigit(s.charAt(left))) {
                left++;
            }

            while (left < right && !Character.isLetterOrDigit(s.charAt(right))) {
                right--;
            }

            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }

        return true;
    }
}

File: neetcode/125. Valid Palindrome/125.kt
/*
Method 1.0: Two pointers
TC: O(N)
SC: O(1)
 */
class Solution {
    fun isPalindrome(s: String): Boolean {
        var left = 0
        var right = s.length - 1

        while (left < right) {
            val a = s[left]
            val b = s[right]

            if (!a.isLetterOrDigit()) {
                left++
            } else if (!b.isLetterOrDigit()) {
                right--
            } else {
                val lowerA = a.lowercaseChar()
                val lowerB = b.lowercaseChar()

                if (lowerA != lowerB) {
                    return false
                } else {
                    left++
                    right--
                }
            }
        }

        return true
    }
}
